### ISSUE TITLE
Feature Request: Completion Item Ranking

### ISSUE THREAD

**Issue by YilunAllenChen** [astral-sh/ty#2927]
### Question

I think a lot of times we'd prefer standard lib typing packages like `typing`, `collections`, `abc` and `dataclasses` over the re-exported packages from third-party packages.

I found this https://github.com/astral-sh/ruff/pull/22460 which suggests that ranking third party over standard libs is a deliberate choice, and I generally agree for non-typing constructs (if you're working in an overlaying framework you should get higher priority on that!). However, I think in a lot of cases, for type-related constructs like `Sequence`, we'd want to use std lib ones, and they're ranked unfavorably low.

See screenshot - `Sequence` from `typing` and `collections.abc` are ranked lower than a bunch of third-party re-exports
<img width="948" height="399" alt="Image" src="https://github.com/user-attachments/assets/8798490f-e406-44a2-bbdc-cd7d80157eb3" />

I'm proposing that we include `typing`, `typing_extensions`, `collections`, `abc` and (perhaps a bit too opinionated) `dataclasses` to the list of special standard libs that **rank higher** than third party dependencies. Other standard libs like `os`, `sys` can remain ranked lower than third party dep as they actually contain implementations of things and hence are more likely to be overlayed when working with some sort of framework.

Alternatively if we want to be super flexible we can also try to make it configurable via settings like

```toml
[tool.uv]
prefer_imports_from = ['abc', 'collections']  # I'm bad at naming, but you get the point
```

### Version

ty 0.0.19

**Comment by dhruvmanila** [comment]
For reference, here's the discussion about this ordering of prioritizing third-party library over stdlib: https://github.com/astral-sh/ruff/pull/22460#discussion_r2675369983

cc @BurntSushi for more context but I think I agree that we should prioritize symbols from (at least) `typing` than third-party symbols. An alternative here is that ty detects that the user is in an annotation position which means that `typing` / `typing_extensions` should be preferable over third-party symbols regardless.

I see you already have a PR up but it would be useful to reach a consensus first before updating the ordering.

> I'm proposing that we include `typing`, `typing_extensions`, `collections`, `abc` and (perhaps a bit too opinionated) `dataclasses` to the list of special standard libs that **rank higher** than third party dependencies.

I think I'd prefer to keep this scoped to `typing` for now and we can think about whether `collections` and `abc` should be included as well but I'm curious to know more about why do you want to include them.

**Comment by sinon** [comment]
> > I'm proposing that we include `typing`, `typing_extensions`, `collections`, `abc` and (perhaps a bit too opinionated) `dataclasses` to the list of special standard libs that **rank higher** than third party dependencies.
> 
> I think I'd prefer to keep this scoped to `typing` for now and we can think about whether `collections` and `abc` should be included as well but I'm curious to know more about why do you want to include them.

`collections.abc` makes sense to include as it falls into the same bucket as `typing` as it's where most of the mature typing containers moved to from `typing` in 3.9/3.10 (I think they are all still in `typing` but they are deprecated aliases to the `collections.abc`). `ruff` has the `pyupgrade` rules which will translate these, so it would be an odd UX for `ty` to promote `typing.Sequence` for `ruff` to then run and replace that with `abc.collections.Sequence` 🤔 

https://docs.python.org/3/library/typing.html#typing.Sequence
https://docs.python.org/3/library/collections.abc.html#collections.abc.Sequence
https://docs.astral.sh/ruff/rules/deprecated-import/

EDIT
Though maybe the nicer solution is for `ty` LSP could use the fact one is deprecated to down rank (assuming this data is transparently available from typeshed/stdlib)

**Comment by YilunAllenChen** [comment]
Thank you for the response! Yup the PR is more provisional than anything, happy to reach consensus here before committing more code.

Totally agree with sinon - the container is exactly the argument why I included collections in the first place. So I think having typing and collections at least makes sense because its consistent with ruffs behavior too.

abc and data classes might be a bit too opinionated, so happy to leave them out. I'm now also a bit hesitant about typing_extensions because I realized that it's not a part of standard lib...

**Comment by BurntSushi** [comment]
Clarifying question: what does your `pyproject.toml` look like for the example you gave? i.e., Does it include jedi, pyasn1 and sqlalchemy? If not, it's possible that https://github.com/astral-sh/ty/issues/2366 would help here. But I suppose having those as direct dependencies is likely pretty common, so it's probably not something to rely upon.

In https://github.com/astral-sh/ruff/pull/22460#discussion_r2676446460 I'm pretty open to changing things here. Starting with `typing` and `collections` seems like it would make sense to me. In particular, both are incredibly common and I don't _think_ there are commonly drop-in third party replacements for them? (Unlike `regex` for `re`, for example.)

> Though maybe the nicer solution is for `ty` LSP could use the fact one is deprecated to down rank (assuming this data is transparently available from typeshed/stdlib)

This should already be happening: https://github.com/astral-sh/ruff/blob/f8d1d29ec7f1ff0da7e0272c8388036f35e815b7/crates/ty_ide/src/completion.rs#L1226-L1225

Although, it is the lowest criteria used. So the criteria that sorts based on what "dependency kind" the module is will override that. So I think the deprecation criterion should probably be moved above that.

**Comment by YilunAllenChen** [comment]
> Does it include jedi, pyasn1 and sqlalchemy? 

It does not. #2366 looks like it could definitely help; though like you said, maybe it won't work consistently for everything that we'd want.

> In particular, both are incredibly common and I don't think there are commonly drop-in third party replacements for them? 

Yeah exactly, and I'd argue that there never will / need ot be replacements, because these are just type annotations (no implementation, so why bother replacing them?)

downranking deprecated things also makes a lot of sense to me, happy to move that above dependency kind if we think it makes sense to lump into the same PR

**Comment by BurntSushi** [comment]
@YilunAllenChen Yeah that seems reasonable to me!

**Comment by marcofrasvda** [comment]
Can't wait for this, the default sorting is not very helpful and caused me to wonder why the following code did not show type errors:
```py
class Attribute(TypedDict):
    force_report: Literal[True]
    
data: Attribute = {
    'force_report': False,
}
```

It turns out it happened because by habit while writing `Literal` I selected the first item in the suggestions list, and instead of `typing.Literal` it imported `pygments.Literal`.
